src/visor/blockdb/unspent.go

```diff
@@ -0,0 +1,297 @@
+package blockdb
+
+import (
+	"fmt"
+
+	"github.com/boltdb/bolt"
+	"github.com/skycoin/skycoin/src/cipher"
+	"github.com/skycoin/skycoin/src/cipher/encoder"
+	"github.com/skycoin/skycoin/src/coin"
+	"github.com/skycoin/skycoin/src/visor/bucket"
+)
+
+var (
+	xorhashKey = []byte("xorhash")
+)
+
+// UnspentPool unspent outputs pool
+type UnspentPool struct {
+	db   *bolt.DB
+	pool *bucket.Bucket
+	meta *bucket.Bucket
+}
+
+type unspentMeta struct {
+	*bolt.Bucket
+}
+
+func (m unspentMeta) getXorHash() (cipher.SHA256, error) {
+	if v := m.Get(xorhashKey); v != nil {
+		var hash cipher.SHA256
+		copy(hash[:], v[:])
+		return hash, nil
+	}
+
+	return cipher.SHA256{}, nil
+}
+
+func (m *unspentMeta) setXorHash(hash cipher.SHA256) error {
+	return m.Put(xorhashKey, hash[:])
+}
+
+type uxOut struct {
+	*bolt.Bucket
+}
+
+func (uo uxOut) get(hash cipher.SHA256) (*coin.UxOut, bool, error) {
+	if v := uo.Get(hash[:]); v != nil {
+		var out coin.UxOut
+		if err := encoder.DeserializeRaw(v, &out); err != nil {
+			return nil, false, err
+		}
+		return &out, true, nil
+	}
+	return nil, false, nil
+}
+
+func (uo uxOut) set(hash cipher.SHA256, ux coin.UxOut) error {
+	v := encoder.Serialize(ux)
+	return uo.Put(hash[:], v)
+}
+
+func (uo *uxOut) delete(hash cipher.SHA256) error {
+	return uo.Delete(hash[:])
+}
+
+// NewUnspentPool creates new unspent pool instance
+func NewUnspentPool(db *bolt.DB) (*UnspentPool, error) {
+	up := &UnspentPool{db: db}
+
+	pool, err := bucket.New([]byte("unspent_pool"), db)
+	if err != nil {
+		return nil, err
+	}
+	up.pool = pool
+
+	meta, err := bucket.New([]byte("unspent_meta"), db)
+	if err != nil {
+		return nil, err
+	}
+	up.meta = meta
+
+	return up, nil
+}
+
+// resetWithTx resets the unspent pool buckets
+func (up *UnspentPool) resetWithTx(tx *bolt.Tx) error {
+	if err := tx.DeleteBucket(up.pool.Name); err != nil {
+		return err
+	}
+
+	_, err := tx.CreateBucket(up.pool.Name)
+	if err != nil {
+		return err
+	}
+
+	if err := tx.DeleteBucket(up.meta.Name); err != nil {
+		return err
+	}
+
+	_, err = tx.CreateBucket(up.meta.Name)
+	return err
+}
+
+func (up *UnspentPool) addWithTx(tx *bolt.Tx, ux coin.UxOut) (err error) {
+	// all updates will rollback if return is not nil
+	// in case of unexpected panic, we must catch it and return error
```

Comment:
The ux.Hash may panic, actually lots of function in encoder package might panic and it's function name does not have a `Must` prefix.